addons/onedrive/models.py

```diff
@@ -0,0 +1,283 @@
+                                BaseStorageAddon)
+from addons.onedrive import settings
+from addons.onedrive.client import OneDriveClient
+from addons.onedrive.settings import DEFAULT_ROOT_ID
+from addons.onedrive.serializer import OneDriveSerializer
+from framework.auth import Auth
+from framework.exceptions import HTTPError
+from osf.models.external import ExternalProvider
+from osf.models.files import File, Folder, BaseFileNode
+from website.util import api_v2_url
+
+logger = logging.getLogger(__name__)
+
+
+class OneDriveFileNode(BaseFileNode):
+    _provider = 'onedrive'
+
+
+class OneDriveFolder(OneDriveFileNode, Folder):
+    pass
+
+
+class OneDriveFile(OneDriveFileNode, File):
+    pass
+
+
+class OneDriveProvider(ExternalProvider):
+    name = 'Microsoft OneDrive'
+    short_name = 'onedrive'
+
+    client_id = settings.ONEDRIVE_KEY
+    client_secret = settings.ONEDRIVE_SECRET
+
+    auth_url_base = settings.ONEDRIVE_OAUTH_AUTH_ENDPOINT
+    callback_url = settings.ONEDRIVE_OAUTH_TOKEN_ENDPOINT
+    auto_refresh_url = settings.ONEDRIVE_OAUTH_TOKEN_ENDPOINT
+    default_scopes = ['wl.basic wl.signin onedrive.readwrite wl.offline_access']
+
+    refresh_time = settings.REFRESH_TIME
+
+    _drive_client = OneDriveClient()
+
+    def handle_callback(self, response):
+        """View called when the Oauth flow is completed. Adds a new OneDriveUserSettings
+        record to the user and saves the user's access token and account info.
+        """
+        user_info = self._drive_client.user_info_for_token(response['access_token'])
+
+        return {
+            'provider_id': user_info['id'],
+            'display_name': user_info['name'],
+            'profile_url': user_info['link']
+        }
+
+    def fetch_access_token(self, force_refresh=False):
+        self.refresh_oauth_key(force=force_refresh)
+        return self.account.oauth_key
+
+
+class UserSettings(BaseOAuthUserSettings):
+    """Stores user-specific onedrive information
+    """
+    oauth_provider = OneDriveProvider
+    serializer = OneDriveSerializer
+
+
+class NodeSettings(BaseOAuthNodeSettings, BaseStorageAddon):
+    """Individual OneDrive settings for a particular node.
+
+    QUIRKS::
+
+    * OneDrive personal and OneDrive for Business users will have only one drive
+      available.  This addon is built around this assumption.  Users using this with
+      a SharePoint team site will have several other drives available, but this use
+      case is not supported or tested.  See:
+      https://dev.onedrive.com/drives/list-drives.htm#remarks
+
+    * OneDrive is an ID-based provider like Box. The identifier for the root folder
+      is defined in the settings.
+
+    """
+    oauth_provider = OneDriveProvider
+    serializer = OneDriveSerializer
+
+    folder_id = models.TextField(null=True, blank=True)
+    folder_path = models.TextField(null=True, blank=True)
+    user_settings = models.ForeignKey(UserSettings, null=True, blank=True)
+
+    _api = None
```

Comment:
Shouldn't this be an instance attribute rather than a class attribute? Wouldn't `OneDriveProvider` instances be shared if this is a class attribute?